src/hotspot/share/opto/vectorIntrinsics.cpp

```diff
@@ -1382,68 +1382,78 @@ bool LibraryCallKit::inline_vector_mem_masked_operation(bool is_store) {
 }
 
-// <C,
-//  V extends Vector<?>,
-//  W extends Vector<Integer>,
-//  S extends VectorSpecies<E>,
-//  M extends VectorMask<E>,
-//  E>
-// V loadWithMap(Class<? extends V> vectorClass, Class<M> maskClass, Class<E> elementType, int length,
-//               Class<? extends Vector<Integer>> vectorIndexClass,
-//               Object base, long offset, // Unsafe addressing
-//               W index_vector, M m,
-//               C container, int index, int[] indexMap, int indexM, S s, // Arguments for default implementation
-//               LoadVectorOperationWithMap<C, V, E, S, M> defaultImpl)
+//  <C,
+//   V extends Vector<E>,
+//   IV extends Vector<IE>,
+//   S extends VectorSpecies<E>,
+//   M extends VectorMask<E>,
+//   E, IE>
+//  V loadWithMap(Class<? extends V> vClass, Class<M> mClass, Class<E> eClass,
+//                int length, Class<? extends IV> ivClass, Class<IE> ieClass,
+//                Object base, long baseOffset, // Unsafe addressing
+//                IV indexMap, long scale, M m,
+//                C container, S s, // Arguments for default implementation
+//                LoadVectorOperationWithMap<C, V, IV, S, M, E, IE> defaultImpl)
 //
 //  <C,
 //   V extends Vector<E>,
-//   W extends Vector<Integer>,
+//   IV extends Vector<IE>,
 //   M extends VectorMask<E>,
-//   E>
-//  void storeWithMap(Class<? extends V> vectorClass, Class<M> maskClass, Class<E> elementType,
-//                    int length, Class<? extends Vector<Integer>> vectorIndexClass, Object base, long offset,    // Unsafe addressing
-//                    W index_vector, V v, M m,
-//                    C container, int index, int[] indexMap, int indexM, // Arguments for default implementation
-//                    StoreVectorOperationWithMap<C, V, M, E> defaultImpl)
+//   E, IE>
+//  void storeWithMap(Class<? extends V> vClass, Class<M> mClass, Class<E> eClass,
+//                    int length, Class<? extends IV> ivClass, Class<IE> ieClass,
+//                    Object base, long baseOffset, // Unsafe addressing
+//                    IV indexMap, long scale, M m, V v,
+//                    C container, // Arguments for default implementation
+//                    StoreVectorOperationWithMap<C, V, IV, M, E, IE> defaultImpl)
 //
 bool LibraryCallKit::inline_vector_gather_scatter(bool is_scatter) {
   const TypeInstPtr* vector_klass     = gvn().type(argument(0))->isa_instptr();
   const TypeInstPtr* mask_klass       = gvn().type(argument(1))->isa_instptr();
   const TypeInstPtr* elem_klass       = gvn().type(argument(2))->isa_instptr();
   const TypeInt*     vlen             = gvn().type(argument(3))->isa_int();
-  const TypeInstPtr* vector_idx_klass = gvn().type(argument(4))->isa_instptr();
+  const TypeInstPtr* idx_vector_klass = gvn().type(argument(4))->isa_instptr();
+  const TypeInstPtr* idx_elem_klass   = gvn().type(argument(5))->isa_instptr();
+  const TypeLong*    scale_type       = gvn().type(argument(10))->isa_long();
```

Comment:
Can we caculate the scale value from the `elem_bt` instead of adding such a constant argument?

Reply:
This is a provision for support of memory segment access where offsets are not scaled.

Reply:
Got it. Thanks!